gemelli/preprocessing.py

```diff
@@ -7,12 +7,267 @@
+import io
+import sys
+import os
+import t2t.nlevel as nl
 import numpy as np
+import pandas as pd
 from biom import Table
+from skbio import TreeNode
 from .base import _BaseConstruct
+from q2_types.tree import NewickFormat
+from gemelli._defaults import DEFAULT_MTD
+from skbio.diversity._util import _vectorize_counts_and_tree
+from bp import parse_newick, to_skbio_treenode
 
 
-def tensor_rclr(T):
+VALID_TAXONOMY_COLUMN_NAMES = ('taxon', 'taxonomy')
+
+
+class TaxonomyError(Exception):
+    pass
+
+
+def create_taxonomy_metadata(phylogeny, traversed_taxonomy=None):
+    """
+        Create a pd.DataFrame with index 'Feature ID' and column 'Taxon'.
+
+        This method will traverse phylogeny using TreeNode.traverse() and
+        use the node names found in phylogeny as the index 'Feature ID'.
+        Thus, traversed_taxonomy must be parallel List of phylogeny (i.e. a
+        List that layes out the taxonomy in a traverse() fashion).
+
+        Parameters
+        ----------
+        phylogeny: TreeNode
+        traversed_taxonomy: List
+    """
+    f_id = []
+    tax = []
+    if traversed_taxonomy is not None:
+        # add taxonomy for all nodes in tree
+        for i, node in enumerate(phylogeny.traverse(include_self=True)):
+            # metadata.append([node.name, traversed_taxonomy[i]])
+            f_id.append(node.name)
+            tax.append(traversed_taxonomy[i])
+    else:
+        # create empty dataframe
+        f_id = ['None']
+        tax = ['None']
+    returned_taxonomy = pd.DataFrame(data={'Feature ID': f_id, 'Taxon': tax})
+    returned_taxonomy.set_index(keys='Feature ID', inplace=True)
+    return returned_taxonomy
+
+
+def retrieve_t2t_taxonomy(phylogeny, taxonomy=None):
+    """
+    Returns a List containing the taxonomy of all nodes in the tree stored in
+    TreeNode.traverse() order. If taxonomy is None, then None will be returned
+
+    based on :
+    https://github.com/biocore/tax2tree/blob/9b3814fb19e935c06a31e61e848d0f91bcecb305/scripts/t2t#L46
+
+    Parameters
+    ----------
+    phylogeny: TreeNode
+    taxonomy : pd.DataFrame with Index 'Feature ID' and contains a column
+        'taxon' or 'taxonomy' (case insensitive)
+    """
+    if taxonomy is None:
+        # return empty taxonomy
+        return None
+
+    # make copy of phylogeny
+    consensus_tree = phylogeny.copy()
+    # validate and convert taxonomy into a StringIO stream
+    consensus_map = _get_taxonomy_io_stream(taxonomy)
+
+    tipname_map = nl.load_consensus_map(consensus_map, False)
+    tree_ = nl.load_tree(consensus_tree, tipname_map)
+    counts = nl.collect_names_at_ranks_counts(tree_)
+
+    nl.decorate_ntips(tree_)
+    nl.decorate_name_relative_freqs(tree_, counts, 2)
+    nl.set_ranksafe(tree_)
+    nl.pick_names(tree_)
+
+    nl.set_preliminary_name_and_rank(tree_)
+
+    contree, contree_lookup = nl.make_consensus_tree(tipname_map.values())
+    # Disable print statements
+    sys.stdout = open(os.devnull, 'w')
```

Comment:
Not sure if this is strictly required but I'm wary of sole `open` statements. Might be better to encapsulate suppression of console output to a context manager.

Reply:
Fixed. Thanks!